### PR TITLE
Feat: Utilities for doing client-side load-balancing based-on endpoints

### PR DESCRIPTION
**Pull request by yue9944882**

### CODE REVIEW

File: extended/src/main/java/io/kubernetes/client/extended/network/RoundRobinLoadBalanceStrategy.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  private final ReentrantLock reentrantLock = new ReentrantLock();
+
+  private int robinIndex = 0;
+
+  @Override
+  public String chooseIP(List<String> availableIPs) {
+    if (availableIPs == null || availableIPs.size() == 0) {
+      throw new IllegalArgumentException("failed choosing IP target: empty candidates");
+    }
+    reentrantLock.lock();
```
Comment (brendandburns) [inline extended/src/main/java/io/kubernetes/client/extended/network/RoundRobinLoadBalanceStrategy.java:29]: Would it be better to use `AtomicInteger` this lock seems heavy-weight.

Comment (brendandburns) [reply extended/src/main/java/io/kubernetes/client/extended/network/RoundRobinLoadBalanceStrategy.java:29]: Also, though this is "round robin" I think you could actually just choose randomly and you'd get the same behavior, I don't think that round robin requires a deterministic order.

**Comment by brendandburns** [comment]
A few comments on the round robin implementation. Also, probably worth adding a 'todo' to add a supplier that uses an Informer instead of polling.

File: extended/src/main/java/io/kubernetes/client/extended/network/RoundRobinLoadBalanceStrategy.java (hunk 1 above)
Comment (yue9944882) [reply extended/src/main/java/io/kubernetes/client/extended/network/RoundRobinLoadBalanceStrategy.java]: > Would it be better to use AtomicInteger this lock seems heavy-weight.

updated

> Also, though this is "round robin" I think you could actually just choose randomly and you'd get the same behavior, I don't think that round robin requires a deterministic order.

i presume the randomized strategy you're saying is a bit different from the proposing one here. for the cases where the client is raising requests at a lower rate e.g. "watching" requests in one loop. the proposed "round robin" in this pull will make sure the even distribution in order, while if we're doing a randomized "round robin", there's a chance of uneven distribution for lower-rate client, is it?

**Comment by yue9944882** [comment]
> A few comments on the round robin implementation. Also, probably worth adding a 'todo' to add a supplier that uses an Informer instead of polling.

i added a new static helper to adapt the informer into the loadbalancer, PTAL

**Pull request merged**